app/javascript/src/LoginPage/loginForms/SignupForm.jsx

```diff
@@ -53,68 +54,85 @@ const InputFormGroup = (props) => {
 }
 
 class SignupForm extends Component<Props, State> {
-  constructor (props: Props) {
-    super(props)
-    this.state = {
-      username: this.props.user.username,
-      emailAddress: this.props.user.email,
-      firstname: this.props.user.firstname,
-      lastname: this.props.user.lastname,
-      password: '',
-      passwordConfirmation: '',
-      isFilledUsername: undefined,
-      isFilledEmailAddress: undefined,
-      isFilledFirstname: true,
-      isFilledLastname: true,
-      isFilledPassword: undefined,
-      isFilledPasswordConfirmation: undefined
-    }
+  state = {
+    'user[username]': this.props.user.username,
+    'user[email]': this.props.user.email,
+    'user[first_name]': this.props.user.firstname,
+    'user[last_name]': this.props.user.lastname,
+    'user[password]': '',
+    'user[password_confirmation]': '',
+    validation: {}
   }
 
-  handleTextInputUsername = (username: string) =>
-    this.setState({ username, isFilledUsername: username !== '' })
-
-  handleTextInputEmail = (emailAddress: string) =>
-    this.setState({ emailAddress, isFilledEmailAddress: emailAddress !== '' })
-
-  handleTextInputFirstname = (firstname: string) => this.setState({ firstname })
-
-  handleTextInputLastname = (lastname: string) => this.setState({ lastname })
+  handleInputChange = (value: string, event: SyntheticEvent<HTMLInputElement>) => {
+    const isValid = event.currentTarget.required ? validateSingleField(event) : true
+    const validation = {
+      ...this.state.validation,
+      [event.currentTarget.name]: isValid
+    }
 
-  handleTextInputPassword = (password: string) =>
     this.setState({
-      password,
-      isFilledPassword: password !== '',
-      isFilledPasswordConfirmation: this.state.passwordConfirmation === password
+      [event.currentTarget.name]: value,
+      validation
     })
+  }
 
-  handleTextInputPasswordConfirmation = (passwordConfirmation: string) =>
-    this.setState({
-      passwordConfirmation,
-      isFilledPasswordConfirmation: passwordConfirmation !== '' && passwordConfirmation === this.state.password
-    })
+  validateForm = (event: SyntheticEvent<HTMLInputElement>) => {
```

Comment:
This should be extracted to a common one, it's always the same for all the forms... "setState(...) unless invalidFields"...

Reply:
I'm not sure if that's a good idea.
If we extract it, we'd need to use some kind of functional `setState` call. And that's  overkilling. 
If we didn't have a call to `setState` here, I'd say yes, but if we have to pass setState a function just to do the same it's currently doing I don't see the benefit.

Reply:
Ok, so let's see if how do you like this:

1. Have just one form component. You'll use it for those 3 forms.
2. This form will accept the "validators" and "inputs" list as props.  You'll compose the form and validate given those props injected.
3. Profit!

Reply:
Another approach:

1. Keep these forms
2. Compose the validation object and create a validate instance, you could inject this with props
3. Then something like this:
```
(!!validate(...).length) && e.preventDefault(); setState(...)
``` 

OR

Just leave it to single input validation and disable the submit button while invalid/pristine

Reply:
I will go with the simplest approach: single input validation and disabling submit button by default